ocaml/xapi/xapi_vdi_helpers.ml

```diff
@@ -23,6 +23,87 @@ open Threadext
 module D=Debug.Make(struct let name="xapi" end)
 open D
 
+(* Attach the rrd-stats vdi to dom0 and get the device *)
+let get_rrd_vdi_device ~__context ~vdi =
+	let open Db_filter_types in
+	let vbds = Db.VBD.get_refs_where ~__context ~expr:(And (
+		Eq (Field "VDI", Literal (Ref.string_of vdi)),
+		Eq (Field "currently_attached", Literal "true"))
+	) in
+	let (vbd, device) =
+		match vbds with
+		| vbd :: _ -> (vbd, Db.VBD.get_device ~__context ~self:vbd)
+		| [] -> begin
+			let dom0 = Helpers.get_domain_zero ~__context in
+			let vbd = Helpers.call_api_functions ~__context (fun rpc session_id ->
+				Client.VBD.create ~rpc ~session_id ~vM:dom0 ~vDI:vdi ~mode:`RW ~_type:`Disk
+				~empty:false ~userdevice:"autodetect" ~bootable:false ~unpluggable:true
+				~qos_algorithm_type:"" ~qos_algorithm_params:[] ~other_config:[] )
+			in
+			Helpers.call_api_functions ~__context (fun rpc session_id ->
+				Client.VBD.plug ~rpc ~session_id ~self:vbd);
+			(vbd, Db.VBD.get_device ~__context ~self:vbd)
+		end
+	in
+	(vbd, device)
+
+(* Detach the rrd-stats vdi *)
+let detach_rrd_vdi ~__context ~vbd =
+	Helpers.call_api_functions ~__context (fun rpc session_id ->
+		Client.VBD.unplug ~rpc ~session_id ~self:vbd);
+	Helpers.call_api_functions ~__context (fun rpc session_id ->
+		Client.VBD.destroy ~rpc ~session_id ~self:vbd)
+
+(* Construct the CStruct for the rrd-stats vdi *)
+let open_rrd_file ~device ~mode =
```

Comment:
I don't think there's a lot of benefit in calling get_rrd_vdi_device and then always immediatly passing the device into open_rrd_file.   If you had open_rrd_file call get_rrd_vdi_device, and changed get_rrd_vdi_device to return a variant which indicated whether the device was fresh or pre-existed, you could then move the magic number checking and formatting logic into here and get it out of the accessors.